docs/source/working_with_aiida/caching.rst

```diff
@@ -65,88 +113,69 @@ This means that caching is enabled for ``TemplatereplacerCalculation`` and ``Str
-
-.. _caching_provenance:
-
-Caching and the Provenance Graph
---------------------------------
-
-The goal of the caching mechanism is to speed up AiiDA calculations by re-using duplicate calculations. However, the resulting provenance graph should be exactly the same as if caching was disabled. This has important consequences on the kind of caching operations that are possible.
-
-The provenance graph consists of nodes describing data, calculations and workchains, and links describing the relationship between these nodes. We have seen that the hash of a node is used to determine whether two nodes are equivalent. To successfully use a cached node however, we also need to know how the new node should be linked to its parents and children.
-
-In the case of a plain data node, this is simple: Copying a data node from an equivalent node should not change its links, so we just need to preserve the links which this new node already has.
-
-For calculations, the situation is a bit more complex: The node can have inputs and creates new data nodes as outputs. Again, the new node needs to keep its existing links. For the outputs, the calculation needs to create a copy of each node and link these as its outputs. This makes it look as if the calculation had produced these outputs itself, without caching.
+3. Run your calculation again. The node in questions should no longer be reused.
 
-Finally, workchains can create links not only to nodes which they create themselves, but also to nodes created by a calculation that they called, or even their ancestors. This is where caching becomes impossible. Consider the following example (using workfunctions for simplicity):
 
-.. code:: python
 
-    from aiida.engine import workfunction
-    from aiida.orm import Int
+.. _caching_limitations:
 
-    @workfunction
-    def select(a, b):
-        return b
+Limitations
+-----------
 
-    d = Int(1)
-    r1 = select(d, d)
-    r2 = select(Int(1), Int(1))
+1. The current implementation of caching for data nodes clones not only the graph representation of the reused node but also the underlying data in the database and file repository. 
+   This could be improved in order to reduce data duplication.
 
-The two ``select`` workfunctions have the same inputs as far as their hashes go. However, the first call uses the same input node twice, while the second one has two different inputs. If the second call should be cached from the first one, it is not clear which of the two input nodes should be returned.
+2. The caching mechanism for calculations *should* trigger only when the inputs and the calculation to be performed are exactly the same.  
+   Edge cases where this assumption might be violated include cases where the calculation parser is in a different python module than the calculation and the developer made changes without updating the version number of the plugin.
```

Comment:
Made changes to the calculation or the parser? It is even worse because as for as I know, for `CalcJob` only the `parser_name` is stored in the attributes and so that's the only thing that is included in computing the hash. Even if the parser version *was* updated properly, this will not be detected by the hashing I think

Reply:
what about this:
https://github.com/aiidateam/aiida-core/blob/56e0100c8098796f8a9f8ed120be06d8f5f29b97/aiida/orm/nodes/process/calculation/calcjob.py#L106-L117

Reply:
That only includes the version of the module that contains the calculation job class though. If the parser lives in a completely different module, or it lives in the same, but its implementation was changed without the version being adapted, it will not be caught.